### DISCUSSION TITLE
Typed loader params with TS generics

### DISCUSSION THREAD

**Discussion by nik-webdevelop** [remix-run/react-router#11002] [Proposals]
It would be really nice to have the ability to strictly type loader parameters. As an example, it could be something like this: 

```tsx
const USER_ROUTE = '/user/:userID';

export const loader = ({ params }: LoaderFunctionArgs<typeof USER_ROUTE>) => {
  const id = params.userID;
}
```

instead of typecasting all the time or jumping through hoops with self-made wrappers that solve only type problems

```tsx
import { ParamParseKey } from 'react-router-dom';
const USER_ROUTE = '/user/:userID';

export const loader = ({ params }: LoaderFunctionArgs) => {
  type TypedParams = Record<ParamParseKey<typeof USER_ROUTE>, string>;
  const id = (params as TypedParams).userID;
}
```

https://github.com/remix-run/react-router/discussions/10985#discussioncomment-7472278

**Comment by akamfoad** [comment]
Implemented this to have the same behavior as we love it in Remix, except the serialization since RR will pass us the same data we returned from the Loader and Actions.

Check out #11021
